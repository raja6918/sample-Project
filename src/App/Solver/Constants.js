import { evaluateRegex } from '../Data/utils/utils.js';

export const NAME_REGEX = '^[a-zA-Z0-9].*';

export const SOLVER_FORM = 'DATA.solver.form';
export const SOLVER_GENERAL = `${SOLVER_FORM}.section.general`;

export const SOLVER_FIELDS = [
  'solverTask',
  'crewGroup',
  'rule',
  'name',
  'description',
  'recipeNames',
];

export const REQUIRED_FIELDS = [
  'solverTask',
  'crewGroup',
  'rule',
  'recipeNames',
  'name',
];
export const POLLING_INTERVAL = 5000;

export const SOLVER_RUNNING_STATUS = [
  'Creating',
  'Waiting',
  'Sending',
  'Running',
  'Fetching',
];
export const CREW_GROUP_RULESET_FAIL_STATUS = [
  'Creating',
  'Launching',
  'READY_TO_LAUNCH',
  'Waiting',
  'Sending',
  'Running',
  'Fetching',
  'Done-success',
  'Done-stopped',
];
export const isRequiredPredicate = val => val !== '' && val !== null;

export const customPredicates = {
  name: val => evaluateRegex(NAME_REGEX, val),
};
