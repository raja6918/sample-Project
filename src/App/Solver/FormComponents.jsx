import styled from 'styled-components';
import FormControlMUI from '@material-ui/core/FormControl';
import TextField from '@material-ui/core/TextField';

export const FormControl = styled(FormControlMUI)`
  width: 100%;
`;

export const Input = styled(TextField)`
  width: 100%;
  margin: 0;
  &.error > label {
    color: #d10000 !important;
  }
`;

export const GeneralTitle = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: flex-end;
  height: 28px;
`;

export const SpecificationTitle = styled.div`
  padding-left: 7px;
  & h2 {
    margin-top: 15px;
  }
`;

export const SolverFormHeader = styled.span`
  padding-left: 7px;
`;
