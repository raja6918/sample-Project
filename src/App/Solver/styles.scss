.rcp-grid {
    z-index: 99999999;
    .Select-menu-outer{top: auto; bottom: 100%}

    .delete-icon {
        width: 40;
        height: 40;
        padding: 0;
        position: absolute;
    }
    .drag-handle-cls {
        color: #7e7e7e;
        position: absolute; 
        cursor: move;
        background-color: transparent !important;
    }
}
.rcp-container {
    .recipe-label {
        display:block;
        font-weight: 500;
        font-size: 14px;
        line-height: 16px;
        color: #000000;
        padding-bottom: 10px;
    }
    
    .recipe-desc {
        display:block;
        position: absolute;
        padding-bottom: 20px;
        font-size: 12px;
        font-family: Roboto;
        font-weight: normal;
        color: #000000;
        line-height: 14px;
        width: 321px;
        height: 28px;
    }
    
}
