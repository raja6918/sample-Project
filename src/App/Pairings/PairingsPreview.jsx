import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import { pull } from 'lodash';
import history from '../../history';
import { previewChannel } from '../../utils/broadCast';
import sessionStorage, { localStorage } from '../../utils/storage';
import { savePreview } from '../../services/Solver';
import scenarioService from '../../services/Scenarios';
import PairingsWithErrorHandler from './Pairings';
import ModalLoader from '../../components/ModalLoader';
import Notification from '../../components/Notification';
import withErrorHandler from '../../components/ErrorHandler/withErrorHandler';
import { connect } from 'react-redux';
import { READ_ONLY } from '../../constants';
import { currentScenario } from '../../utils/common';

export class PairingsPreview extends Component {
  constructor(props) {
    super(props);

    this.state = {
      previewDetails: null,
      openLoader: false,
      titleLoader: '',
      snackMessage: null,
      snackType: '',
    };
  }

  clearLocalStorage = () => {
    // Remove openItem from localstorage
    const previewId = this.props.match.params.previewID;
    const openPreviews = localStorage.getItem('openPreviews');

    const filteredPreviews = openPreviews.filter(preview =>
      preview ? preview.previewId !== parseInt(previewId, 10) : false
    );

    localStorage.setItem('openPreviews', filteredPreviews);
  };

  setupBeforeUnloadListener = () => {
    const { previewID } = this.props.match.params;

    const deletedPreviews = localStorage.getItem('deletedPreviews') || [];
    localStorage.setItem('deletedPreviews', [...deletedPreviews, previewID]);

    previewChannel.postMessage({
      title: 'DELETE_ID',
      data: previewID,
    });

    this.clearLocalStorage();
  };

  componentDidMount() {
    const previewId = this.props.match.params.previewID;

    // Register the event listener
    window.addEventListener('beforeunload', this.setupBeforeUnloadListener);

    // grab openPreview from session storage
    const openPreview = sessionStorage.getItem('openPreview');

    const openPreviews = localStorage.getItem('openPreviews') || [];
    let previewDetails = openPreviews.find(preview =>
      preview ? preview.previewId === parseInt(previewId, 10) : false
    );

    // if user refresh the page and local storage content is lost then use session storage content
    if (!previewDetails) {
      previewDetails = openPreview;
      // and then set localstorage
      localStorage.setItem('openPreviews', openPreviews.concat(openPreview));
      // Remove id from deletedPreviews
      const deletedPreviews = localStorage.getItem('deletedPreviews') || [];
      pull(deletedPreviews, previewId);
      localStorage.setItem('deletedPreviews', deletedPreviews);
    }

    this.setState({ previewDetails });

    // check whether preview is readonly
    const openScenario = sessionStorage.getItem('openScenario');
    const readOnly = openScenario.status === READ_ONLY;

    if (readOnly) {
      this.disableSave();
    }

    // set history state
    history.push({
      state: { ...previewDetails, readOnly },
    });

    // If session storage is not set then set it - happen when tab is open first time
    if (!openPreview) {
      sessionStorage.setItem('openPreview', {
        pathname: location.pathname,
        ...previewDetails,
      });
    }
  }

  handleClose = async () => {
    const { userData } = this.props;
    const userId = userData ? userData.id : '';

    // Unregister the event listener
    window.removeEventListener('beforeunload', this.setupBeforeUnloadListener);

    try {
      // call delete scenario API
      await scenarioService.deletePreview(
        parseInt(this.props.match.params.previewID, 10),
        userId,
        true
      );

      this.clearLocalStorage();
      window.close();
    } catch (error) {
      // Re-Register the event listener
      window.addEventListener('beforeunload', this.setupBeforeUnloadListener);
      this.props.reportError({ error });
    }
  };

  handleSave = async () => {
    const crewGroupName = this.state.previewDetails.crewGroups
      ? this.state.previewDetails.crewGroups[0].name
      : '';
    this.openLoader(crewGroupName);
    this.onClearSnackBar();

    const scenario = currentScenario();
    const scenarioId = scenario ? scenario.id : null;

    try {
      // Call save preview API
      const data = await savePreview(
        parseInt(this.props.match.params.previewID, 10),
        scenarioId
      );
      // If save is success without any warning
      if (data.status === 'SUCCESS') {
        this.setState({
          snackMessage: this.props.t('SUCCESS.PAIRING_PREVIEW.SAVE', {
            solverRequestName: this.state.previewDetails.solverName,
            crewGroupName,
          }),
          snackType: 'success',
        });
      }

      // If save is success with warning
      if (data.status === 'SUCCESS_WARNING') {
        this.setState({
          snackMessage: this.props.t('SUCCESS.PAIRING_PREVIEW.SAVE_WARNING', {
            solverRequestName: this.state.previewDetails.solverName,
            crewGroupName,
          }),
          snackType: 'info',
        });
      }

      // Broadcast message to all opened tabs to remove pairingStore from session storage of its parent scenario
      const parentScenario = currentScenario();
      if (parentScenario) {
        previewChannel.postMessage({
          title: 'PREVIEW_SAVED',
          scenarioId: parentScenario.id,
        });
      }

      this.disableSave();
    } catch (error) {
      // If save completely failed with 400 status
      this.setState({
        snackMessage: this.props.t('ERRORS.PAIRING_PREVIEW_SAVE.message'),
        snackType: 'error',
      });
    } finally {
      this.closeLoader();
    }
  };

  disableSave = () => {
    sessionStorage.setItem('disableSave', { status: true });
  };

  openLoader = crewGroupName => {
    this.setState({
      openLoader: true,
      titleLoader: this.props.t('PAIRINGS.solverResults.loader', {
        crewGroupName,
      }),
    });
  };

  closeLoader = () => {
    this.setState({ openLoader: false });
  };

  onClearSnackBar = () => {
    this.setState({
      snackMessage: null,
      snackBarType: '',
    });
  };

  render() {
    return (
      this.state.previewDetails && (
        <Fragment>
          <PairingsWithErrorHandler
            previewMode={true}
            handleClose={this.handleClose}
            handleSave={this.handleSave}
            {...this.props}
          />
          <ModalLoader
            open={this.state.openLoader}
            title={this.state.titleLoader}
            color="white"
          />
          <Notification
            message={this.state.snackMessage}
            clear={this.onClearSnackBar}
            type={this.state.snackType}
            displayCloseButton={this.state.snackType === 'info' ? true : false}
            autoHideDuration={this.state.snackType === 'info' ? 3600000 : 5000}
          />
        </Fragment>
      )
    );
  }
}

PairingsPreview.propTypes = {
  t: PropTypes.func.isRequired,
  match: PropTypes.shape().isRequired,
  reportError: PropTypes.func.isRequired,
};

const mapStateToProps = state => {
  return { userData: state.user.userData };
};
const PairingsPreviewComponent = connect(mapStateToProps)(PairingsPreview);

export default withErrorHandler(PairingsPreviewComponent);
