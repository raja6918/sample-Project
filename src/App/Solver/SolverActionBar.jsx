import React, { Component } from 'react';

import styled from 'styled-components';
import PropTypes from 'prop-types';

import Grid from '@material-ui/core/Grid';
import IconButton from '@material-ui/core/IconButton';
import Icon from '@material-ui/core/Icon';
import history from '../../history';
import GenericTextField from '../../components/GenericTextField';
import ErrorDialog from '../../components/Dialog/ErrorDialog';
import ModalLoader from '../../components/ModalLoader';
import storage, { localStorage } from '../../utils/storage';
import {
  openInNewTab,
  getBrowserSessionId,
  getFilteredCrewBasesandRules,
} from './utils';
import { getPreviewId } from '../../services/Solver/index';
import { connect } from 'react-redux';
import AccessEnabler from '../../components/AccessEnabler';

const ButtonItem = styled(IconButton)`
  z-index: 1;
  overflow: visible;
  height: 48px;
  width: 48px;
  padding: 0px;
  & > span {
    display: block;
    line-height: 0px;
  }
`;

const BoldedGenericTextField = styled.div`
  input {
    font-weight: 400;
    font-size: 20px;
  }
`;

const ActionButtonContainer = styled(Grid)`
  margin-top: 5px;
  & div {
    z-index: 10;
    margin: 0 7px;
  }
  & div:first-child {
    margin-left: -4px;
  }
  & div:last-child {
    margin-right: -10px;
  }
`;
const IconLabel = styled.span`
  font-family: 'Roboto-Regular', sans-serif;
  font-size: 12px;
`;
const BarContainer = styled.div`
  padding: 25px 40px 0 40px;
  & h2 {
    margin: 0 0 20px 0;
    font-size: 20px;
    color: rgba(0, 0, 0, 0.87);
    font-weight: normal;
  }
`;
const ActionButton = ({
  icon,
  label,
  onClick,
  readOnly,
  active,
  isFetching,
}) => {
  const disabled = !active || readOnly || isFetching;
  // if (activeBtns) {
  //   disabled = !activeBtns.includes(status) || readOnly;
  // }

  const color = disabled ? 'disabled' : 'primary';

  return (
    <Grid item xm={2}>
      <ButtonItem onClick={onClick} disabled={disabled}>
        <Icon color={color}>{icon}</Icon>
        <IconLabel>{label}</IconLabel>
      </ButtonItem>
    </Grid>
  );
};

export class SolverActionBar extends Component {
  state = {
    name: '',
    isFetching: false,
    maxPreviewDialogOpen: false,
    active: false,
    popupBlockDialogOpen: false,
    showErrors: this.props.showErrors,
  };

  onTriggerAction = action => {
    this.props.onUpdateState(action);
  };

  onPreview = async () => {
    const { id, scenarioSnapshot } = this.props.activeRequest;
    const { scenarioId, scenarioName } = this.props;
    const openItemId = scenarioId;
    const openItemName = scenarioName;

    // get or create browser session id
    const browserSessionId = await getBrowserSessionId();

    const openPreviews = localStorage.getItem('openPreviews') || [];

    // check max preview limit
    const childPreviews = openPreviews.filter(preview =>
      preview ? preview.browserSessionId === browserSessionId : false
    );
    if (childPreviews.length > 4) {
      this.setState({ maxPreviewDialogOpen: true });
      return;
    }

    // setup Crew Groups and rules
    const { crewGroupName, rulesetName } = this.props.activeRequest;
    let crewGroupId = this.props.crewGroups
      ? this.props.crewGroups.find(
          crewGroup => crewGroup.name === crewGroupName
        )
      : '';
    let rulesetId = this.props.rules
      ? this.props.rules.find(rule => rule.name === rulesetName)
      : '';
    crewGroupId = crewGroupId ? crewGroupId.id : '';
    rulesetId = rulesetId ? rulesetId.id : '';

    const [crewGroups, rules] = getFilteredCrewBasesandRules(
      crewGroupId,
      this.props.crewGroups,
      rulesetId,
      this.props.rules
    );
    let filterPayload = null;
    let timelineLastFilter = null;
    let pairingStore = null;

    try {
      this.setState({ isFetching: true });
      const requestBody = {
        scenarioId: openItemId,
        solverRequestId: id,
        snapShot: scenarioSnapshot,
        crewGroupId: crewGroupId,
        rulesetId: rulesetId,
      };
      const data = await getPreviewId(requestBody);
      if (crewGroupId === '' || !crewGroups.length) {
        crewGroupId = crewGroupName;
        crewGroups.push({ id: crewGroupName, name: crewGroupName });
      }
      if (rulesetId === '' || !rules.length) {
        rulesetId = crewGroupName;
        rules.push({ id: rulesetName, name: rulesetName });
      }

      openPreviews.push({
        previewId: data.previewScenarioId,
        openItemId: data.previewScenarioId,
        openItemName,
        solverId: id,
        browserSessionId,
        crewGroup: crewGroupId,
        crewGroups,
        rule: rulesetId,
        rules,
        solverName: this.props.activeRequest.name,
      });
      localStorage.setItem('openPreviews', openPreviews);

      filterPayload = storage.getItem(`timelineFilter1`);
      storage.removeItem(`timelineFilter1`);
      timelineLastFilter = storage.getItem('timelineLastFilter1');
      storage.removeItem(`timelineLastFilter1`);
      pairingStore = storage.getItem('pairingStore');
      storage.removeItem(`pairingStore`);

      const isOpen = openInNewTab(
        location.protocol +
          '//' +
          location.host +
          '/pairings-preview/' +
          data.previewScenarioId,
        data.previewScenarioId
      );

      if (!isOpen) {
        this.setState({ popupBlockDialogOpen: true });
      }
    } catch (error) {
      if (error.response) this.props.reportError({ error });
    } finally {
      this.setState({ isFetching: false });
      if (filterPayload) {
        storage.setItem(`timelineFilter1`, filterPayload);
      }
      if (timelineLastFilter) {
        storage.setItem(`timelineLastFilter1`, timelineLastFilter);
      }
      if (pairingStore) {
        storage.setItem(`pairingStore`, pairingStore);
      }
    }
  };

  closeMaxPreviewDialog = () => {
    this.setState({ maxPreviewDialogOpen: false });
  };

  closePopupBlockDialog = () => {
    this.setState({ popupBlockDialogOpen: false });
  };

  componentWillMount() {
    this.setState({ name: this.props.activeRequest.name });
  }

  componentWillReceiveProps(nextProps) {
    this.setState({
      name: nextProps.activeRequest.name,
    });
  }

  shouldComponentUpdate() {
    return !this.state.active;
  }

  handleOnBlur = (id, value) => {
    this.setState({ active: false, name: value }, () => {
      if (id && value) {
        this.props.updateSolverName(id, value);
      }
    });
  };
  checkActive = active => {
    this.setState({ active });
  };

  render() {
    const {
      t,
      activeRequest,
      onEndorse,
      readOnly,
      showErrors,
      solverScopes,
    } = this.props;
    const {
      isFetching,
      maxPreviewDialogOpen,
      popupBlockDialogOpen,
      name,
    } = this.state;
    const label = 'SOLVER.actionBar.actionButtons';
    const errorMsg = 'ERRORS.SOLVER';
    const endorseIcon = activeRequest.isEndorsed
      ? 'favorite'
      : 'favorite_border';

    const activePlay =
      activeRequest.status.activeBtns.includes('play') && !showErrors;
    const activeStop = activeRequest.status.activeBtns.includes('stop');
    const activePreview = activeRequest.status.activeBtns.includes('preview');
    const activeEndorse = activeRequest.status.activeBtns.includes('endorse');
    const activeMore = activeRequest.status.activeBtns.includes('more');

    const isDisabled =
      [
        'Done-success',
        'Creating',
        'Waiting',
        'Sending',
        'Running',
        'Fetching',
        'Stopping',
        'Launching',
      ].includes(activeRequest.status.status) || readOnly;
    return (
      <BarContainer>
        <Grid container spacing={5}>
          <Grid item xs={5} xm={5}>
            <AccessEnabler
              scopes={solverScopes.solverManage}
              disableComponent
              render={props => (
                <BoldedGenericTextField>
                  <GenericTextField
                    id="name"
                    value={name}
                    readOnly={isDisabled || props.disableComponent}
                    onBlur={this.handleOnBlur}
                    checkError={true}
                    t={t}
                    errorMsg={`${errorMsg}.name`}
                    isActive={this.checkActive}
                  />
                </BoldedGenericTextField>
              )}
            />
          </Grid>
        </Grid>
        <Grid container spacing={0}>
          <Grid item xs={6} xm={6}>
            <ActionButtonContainer container spacing={0}>
              <AccessEnabler
                scopes={solverScopes.solverManage}
                disableComponent
                render={props => (
                  <ActionButton
                    icon="play_arrow"
                    label={t(`${label}.launch`)}
                    onClick={() => this.onTriggerAction('play')}
                    active={activePlay}
                    readOnly={readOnly || props.disableComponent}
                  />
                )}
              />
              <AccessEnabler
                scopes={solverScopes.solverManage}
                disableComponent
                render={props => (
                  <ActionButton
                    icon="stop"
                    label={t(`${label}.stop`)}
                    onClick={() => this.onTriggerAction('stop')}
                    active={activeStop}
                    readOnly={readOnly || props.disableComponent}
                  />
                )}
              />
              <AccessEnabler
                scopes={solverScopes.solverEvaluate}
                disableComponent
                render={props => (
                  <ActionButton
                    icon="visibility"
                    label={t(`${label}.preview`)}
                    onClick={this.onPreview}
                    active={activePreview}
                    isFetching={isFetching}
                    readOnly={props.disableComponent}
                  />
                )}
              />
            </ActionButtonContainer>
          </Grid>
          <Grid item xs={6} xm={6}>
            <ActionButtonContainer container spacing={0} justify="flex-end">
              <AccessEnabler
                scopes={solverScopes.solverEvaluateOrManage}
                disableComponent
                render={props => (
                  <ActionButton
                    icon={endorseIcon}
                    label={t(`${label}.endorse`)}
                    onClick={onEndorse}
                    active={activeEndorse}
                    readOnly={readOnly || props.disableComponent}
                  />
                )}
              />
              {/* <ActionButton
                icon="more_vert"
                label={t(`${label}.more`)}
                readOnly={readOnly}
                active={activeMore}
              /> */}
            </ActionButtonContainer>
          </Grid>
        </Grid>
        <ErrorDialog
          open={maxPreviewDialogOpen}
          handleOk={this.closeMaxPreviewDialog}
          onExited={this.closeMaxPreviewDialog}
          okText={t('GLOBAL.form.close')}
          title={t('ERRORS.PAIRING_PREVIEW_ERROR.title')}
          bodyText={t('ERRORS.PAIRING_PREVIEW_ERROR.body')}
        />
        <ErrorDialog
          open={popupBlockDialogOpen}
          handleOk={this.closePopupBlockDialog}
          onExited={this.closePopupBlockDialog}
          okText={t('GLOBAL.form.close')}
          title={t('ERRORS.PAIRING_PREVIEW_BLOCKED.title')}
          bodyText={t('ERRORS.PAIRING_PREVIEW_BLOCKED.body')}
        />
        <ModalLoader
          open={isFetching}
          title={t('DATA.solver.preview.loader')}
          color="white"
        />
      </BarContainer>
    );
  }
}

const mapStatetoProps = state => {
  const { newJobStatus: solver } = state;
  return {
    showErrors: solver.showErrors,
  };
};

ActionButton.propTypes = {
  icon: PropTypes.string.isRequired,
  label: PropTypes.string,
  status: PropTypes.string,
  activeBtns: PropTypes.arrayOf(PropTypes.string),
  onClick: PropTypes.func,
  readOnly: PropTypes.bool,
  isFetching: PropTypes.bool,
};
ActionButton.defaultProps = {
  label: '',
  status: null,
  activeBtns: [],
  onClick: () => {},
  readOnly: false,
  isFetching: false,
};

SolverActionBar.propTypes = {
  activeRequest: PropTypes.shape({
    crewGroupName: PropTypes.string,
    crewGroupId: PropTypes.number.isRequired,
    rulesetId: PropTypes.number.isRequired,
    elapsedTime: PropTypes.string,
    id: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
    isEndorsed: PropTypes.bool.isRequired,
    name: PropTypes.string.isRequired,
    status: PropTypes.shape({
      activeBtns: PropTypes.arrayOf(PropTypes.string).isRequired,
      block: PropTypes.bool.isRequired,
      dateCurrent: PropTypes.bool.isRequired,
      icon: PropTypes.string.isRequired,
      label: PropTypes.string.isRequired,
      routes: PropTypes.arrayOf(PropTypes.number).isRequired,
      showTimer: PropTypes.bool.isRequired,
      status: PropTypes.string.isRequired,
      statusId: PropTypes.number.isRequired,
      textBar: PropTypes.string.isRequired,
      timerType: PropTypes.string.isRequired,
    }).isRequired,
    lastModified: PropTypes.string,
  }).isRequired,
  t: PropTypes.func.isRequired,
  onEndorse: PropTypes.func.isRequired,
  onUpdateState: PropTypes.func.isRequired,
  readOnly: PropTypes.bool.isRequired,
  reportError: PropTypes.func.isRequired,
  crewGroups: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      ruleset: PropTypes.number.isRequired,
    })
  ).isRequired,
  rules: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
    })
  ).isRequired,
  showErrors: PropTypes.bool.isRequired,
  solverScopes: PropTypes.shape([]).isRequired,
};

export default connect(mapStatetoProps)(SolverActionBar);
