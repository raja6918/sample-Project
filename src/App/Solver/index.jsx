import Solver from './Solver';

export default Solver;
