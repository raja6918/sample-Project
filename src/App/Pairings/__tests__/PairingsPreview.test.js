import React from 'react';
import { shallow } from 'enzyme';
import storage, { localStorage } from '../../../utils/storage';
import mockedPreview, { mockScenario } from './mockedPreview.mock';

import * as solverService from '../../../services/Solver';
import scenarioService from '../../../services/Scenarios/index';

import { PairingsPreview } from '../PairingsPreview';

import { user } from '../../../reducers/__mocks__/store';

const minProps = {
  match: {
    params: {
      previewID: mockedPreview.previewId,
    },
  },
  reportError: jest.fn(),
  t: jest.fn(),
  location: {},
  userData: user.userData,
};

beforeEach(() => {
  storage.setItem('openScenario', mockScenario);
});

afterEach(() => {
  storage.clear();
  localStorage.clear();
});

test('PairingsPreview Component should not render Pairings component if both storage is empty', () => {
  const wrapper = shallow(<PairingsPreview {...minProps} />);
  expect(wrapper).toMatchSnapshot();

  // check whether previewDetails state is null
  expect(wrapper.state('previewDetails')).toBe(null);
});

test('PairingsPreview Component should render Pairings component if session storage is not empty', () => {
  // set session storage
  storage.setItem('openPreview', mockedPreview);

  const wrapper = shallow(<PairingsPreview {...minProps} />);
  expect(wrapper).toMatchSnapshot();

  // check whether local storage is set
  const openPreviews = localStorage.getItem('openPreviews') || [];
  expect(openPreviews[0]).toEqual(mockedPreview);

  // check whether previewDetails state is also set
  expect(wrapper.state('previewDetails')).toEqual(mockedPreview);
});

test('PairingsPreview Component should render Pairings component if local storage is not empty', () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);

  const wrapper = shallow(<PairingsPreview {...minProps} />);
  expect(wrapper).toMatchSnapshot();

  // check whether session storage is set
  const openPreview = storage.getItem('openPreview');
  expect(openPreview).toEqual({ ...mockedPreview, pathname: '/' });

  // check whether previewDetails state is also set
  expect(wrapper.state('previewDetails')).toEqual(mockedPreview);
});

test('check localstorage is cleared when beforeunload event is triggered', () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);

  global.dispatchEvent(new Event('beforeunload'));

  // check whether session storage is set
  const openPreviews = localStorage.getItem('openPreviews');
  expect(openPreviews).toEqual([]);
});

test('check whether handleSave function is triggered', () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);
  const wrapper = shallow(<PairingsPreview {...minProps} />);

  wrapper.instance().handleSave = jest.fn();
  wrapper.instance().forceUpdate();

  wrapper.find('ErrorHandler').prop('handleSave')();

  expect(wrapper.instance().handleSave).toHaveBeenCalled();

  wrapper.instance().handleSave.mockClear();
});

test('check whether snackType is set to error if error occured in savePreview', async () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);
  const wrapper = shallow(<PairingsPreview {...minProps} />);

  solverService.savePreview = jest.fn(() => Promise.reject());

  expect(wrapper.state('snackType')).toEqual('');

  await wrapper.instance().handleSave();

  expect(wrapper.state('snackType')).toEqual('error');
});

test('check whether snackType is set to success if savePreview service is success', async () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);
  const wrapper = shallow(<PairingsPreview {...minProps} />);

  solverService.savePreview = jest.fn(() =>
    Promise.resolve({ status: 'SUCCESS', scenarioId: 1015 })
  );

  expect(wrapper.state('snackType')).toEqual('');

  await wrapper.instance().handleSave();

  expect(wrapper.state('snackType')).toEqual('success');

  // check whether session storage is set
  const disableSave = storage.getItem('disableSave');
  expect(disableSave.status).toBe(true);
});

test('check whether snackType is set to info if savePreview service is success with warnings', async () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);
  const wrapper = shallow(<PairingsPreview {...minProps} />);

  solverService.savePreview = jest.fn(() =>
    Promise.resolve({ status: 'SUCCESS_WARNING', scenarioId: 1015 })
  );

  expect(wrapper.state('snackType')).toEqual('');

  await wrapper.instance().handleSave();

  expect(wrapper.state('snackType')).toEqual('info');

  // check whether session storage is set
  const disableSave = storage.getItem('disableSave');
  expect(disableSave.status).toBe(true);
});

test('check whether handleClose function is triggered', () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);
  const wrapper = shallow(<PairingsPreview {...minProps} />);

  wrapper.instance().handleClose = jest.fn();
  wrapper.instance().forceUpdate();

  wrapper.find('ErrorHandler').prop('handleClose')();

  expect(wrapper.instance().handleClose).toHaveBeenCalled();

  wrapper.instance().handleClose.mockClear();
});

test('check whether reportError is called if error occured in deleteScenario', async () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);
  const wrapper = shallow(<PairingsPreview {...minProps} />);

  scenarioService.deletePreview = jest.fn(() => Promise.reject());

  await wrapper.instance().handleClose();

  expect(minProps.reportError).toHaveBeenCalled();
});

test('check whether clearLocalStorage is called if deleteScenario is successfull', async () => {
  // set local storage
  localStorage.setItem('openPreviews', [mockedPreview]);
  const wrapper = shallow(<PairingsPreview {...minProps} />);

  wrapper.instance().clearLocalStorage = jest.fn();
  wrapper.instance().forceUpdate();

  scenarioService.deletePreview = jest.fn(() => Promise.resolve());

  await wrapper.instance().handleClose();

  expect(wrapper.instance().clearLocalStorage).toHaveBeenCalled();

  wrapper.instance().clearLocalStorage.mockClear();
});
