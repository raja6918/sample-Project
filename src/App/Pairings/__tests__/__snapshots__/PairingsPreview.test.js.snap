// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PairingsPreview Component should not render Pairings component if both storage is empty 1`] = `""`;

exports[`PairingsPreview Component should render Pairings component if local storage is not empty 1`] = `
<Fragment>
  <ErrorHandler
    handleClose={[Function]}
    handleSave={[Function]}
    location={Object {}}
    match={
      Object {
        "params": Object {
          "previewID": 6213,
        },
      }
    }
    previewMode={true}
    reportError={[MockFunction]}
    t={[MockFunction]}
    userData={
      Object {
        "email": "admin@kronos.com",
        "firstName": "admin",
        "id": "2",
        "lastName": "admin",
        "role": "Administrator",
        "username": "admin",
      }
    }
  />
  <ModalLoader
    background=""
    color="white"
    loaderRequired={true}
    onClose={[Function]}
    open={false}
    subtitle=""
    title=""
  />
  <Notification
    autoHideDuration={5000}
    bottomheight={0}
    buttonText=""
    clear={[Function]}
    displayCloseButton={false}
    message={null}
    onLinkAction={[Function]}
    renderers={
      Object {
        "paragraph": "span",
      }
    }
    type=""
  />
</Fragment>
`;

exports[`PairingsPreview Component should render Pairings component if session storage is not empty 1`] = `
<Fragment>
  <ErrorHandler
    handleClose={[Function]}
    handleSave={[Function]}
    location={Object {}}
    match={
      Object {
        "params": Object {
          "previewID": 6213,
        },
      }
    }
    previewMode={true}
    reportError={[MockFunction]}
    t={[MockFunction]}
    userData={
      Object {
        "email": "admin@kronos.com",
        "firstName": "admin",
        "id": "2",
        "lastName": "admin",
        "role": "Administrator",
        "username": "admin",
      }
    }
  />
  <ModalLoader
    background=""
    color="white"
    loaderRequired={true}
    onClose={[Function]}
    open={false}
    subtitle=""
    title=""
  />
  <Notification
    autoHideDuration={5000}
    bottomheight={0}
    buttonText=""
    clear={[Function]}
    displayCloseButton={false}
    message={null}
    onLinkAction={[Function]}
    renderers={
      Object {
        "paragraph": "span",
      }
    }
    type=""
  />
</Fragment>
`;
